File: kfcore/src/main/java/com/lq/ren/base/ui/obs/WorkObservable.java
package com.lq.ren.base.ui.obs;

/**
 * An observable that return the last value when add an observer.
 */
public class WorkObservable<T> extends Observable<T> {

    private T lastValue;
    private boolean hasLastValue = false;

    @Override
    public void notifyObservers() {
        boolean hasLast;
        T last;
        synchronized (this) {
            hasLast = hasLastValue;
            last = lastValue;
        }
        if (hasLast) {
            setChanged();
            super.notifyObserver(last);
        }
    }

    @Override
    public void notifyObserver(T data) {
        synchronized (this) {
            hasLastValue = true;
            lastValue = data;
        }
        setChanged();
        super.notifyObserver(data);
    }

    @Override
    public void addObserver(Observer<T> observer) {
        super.addObserver(observer);

        // Just update given observer, not affect other observers.
        if (observer != null) {
            boolean hasLast;
            T last;
            synchronized (this) {
                hasLast = hasLastValue;
                last = lastValue;
            }
            if (hasLast) {
                observer.update(this, last);
            }
        }
    }
}

File: kfcore/src/main/java/com/lq/ren/base/ui/obs/WorkObservable.kt
package com.lq.ren.base.ui.obs

/**
 * An observable that return the last value when add an observer.
 */
class WorkObservable<T> : Observable<T>() {

    private var lastValue: T? = null
    private var hasLastValue = false

    override fun notifyObservers() {
        var hasLast: Boolean
        var last: T?
        synchronized(this) {
            hasLast = hasLastValue
            last = lastValue
        }
        if (hasLast) {
            setChanged()
            super.notifyObserver(last)
        }
    }

    override fun notifyObserver(data: T?) {
        synchronized(this) {
            hasLastValue = true
            lastValue = data
        }
        setChanged()
        super.notifyObserver(data)
    }

    override fun addObserver(observer: Observer<T>?) {
        super.addObserver(observer)

        // Just update given observer, not affect other observers.
        if (observer != null) {
            var hasLast: Boolean
            var last: T?
            synchronized(this) {
                hasLast = hasLastValue
                last = lastValue
            }
            if (hasLast) {
                observer.update(this, last)
            }
        }
    }
}
